pkg/runner/runner.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runner
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/tracer"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/types"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type runnerState string
+
+const (
+	RunnerStateNotStarted runnerState = "not-started"
+	RunnerStateStarted    runnerState = "started"
+	RunnerStateCompleted  runnerState = "completed"
+)
+
+type RunnerOpts struct {
+	tracer.RunOpts
+
+	Timeout    time.Duration
+	context    context.Context
+	ebpfParams map[string]types.EBPFParam
+}
+
+// WithTimeout sets the timeout for the gadget
+func WithTimeout(timeout time.Duration) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.Timeout = timeout
+	}
+}
+
+// WithContext sets the context for the gadget
+// It is useful if one wants to cancel the gadget execution
+func WithContext(ctx context.Context) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.context = ctx
+	}
+}
+
+// WithValidateMetadata sets whether the metadata should be validated against the eBPF program
+func WithValidateMetadata(validateMetadata bool) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.ValidateMetadata = validateMetadata
+	}
+}
+
+// WithPullPolicy sets the pull policy for pulling the gadget image
+func WithPullPolicy(pullPolicy string) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.PullPolicy = pullPolicy
+	}
+}
+
+// WithAuthOpts sets the authentication options when pulling the gadget image
+func WithAuthOpts(authOpts *oci.AuthOptions) func(*RunnerOpts) {
+	return func(r *RunnerOpts) {
+		r.AuthOpts = authOpts
+	}
+}
+
+type Runner struct {
+	eventBuffer
+
+	RunnerOpts RunnerOpts
```

Comment:
Does it need to be exported?
```suggestion
	runnerOpts RunnerOpts
```